File: source/includes/fts-tutorial/complex-fts-query.py
import pymongo
import dns

client = pymongo.MongoClient('<connection-string>')
result = client['sample_mflix']['movies'].aggregate([
    {
        '$search': {
            'compound': {
                'must': [
                    {
                        'text': {
                            'query': [
                                'Hawaii', 'Alaska'
                            ],
                            'path': 'plot'
                        }
                    }, {
                        'regex': {
                            'query': '([0-9]{4})',
                            'path': 'plot',
                            'allowAnalyzedField': True
                        }
                    }
                ],
                'mustNot': [
                    {
                        'text': {
                            'query': [
                                'Comedy', 'Romance'
                            ],
                            'path': 'genres'
                        }
                    }, {
                        'text': {
                            'query': [
                                'Beach', 'Snow'
                            ],
                            'path': 'title'
                        }
                    }
                ]
            }
        }
    }, {
        '$project': {
            'title': 1,
            'plot': 1,
            'genres': 1,
            '_id': 0
        }
    }
])

for i in result:
    print(i)

File: source/includes/fts-tutorial/complex-fts-query.kt
import com.mongodb.client.model.Aggregates.limit
import com.mongodb.client.model.Aggregates.project
import com.mongodb.client.model.Filters.eq
import com.mongodb.client.model.Projections.*
import com.mongodb.kotlin.client.coroutine.MongoClient
import kotlinx.coroutines.runBlocking
import org.bson.Document

fun main() {
    // establish connection and set namespace
    val uri = "<connection-string>"
    val mongoClient = MongoClient.create(uri)
    val database = mongoClient.getDatabase("sample_mflix")
    val collection = database.getCollection<Document>("movies")

    runBlocking {
        // define query
        val agg = Document(
            "must", listOf(
                Document("text", Document("query", listOf("Hawaii", "Alaska"))
                    .append("path", "plot")),
                Document(
                    "regex",
                    Document("query", "([0-9]{4})")
                        .append("path", "plot")
                        .append("allowAnalyzedField", true)
                )
            )
        )
            .append(
                "mustNot", listOf(
                    Document("text", Document("query", listOf("Comedy", "Romance"))
                        .append("path", "genres")),
                    Document("text", Document("query", listOf("Beach", "Snow"))
                        .append("path", "title"))
                )
            )

        // run query and print results
        val resultsFlow = collection.aggregate<Document>(
            listOf(
                eq("\$search", eq("compound", agg)),
                project(fields(
                    excludeId(),
                    include("title", "plot", "genres")
                ))
            )
        )
        resultsFlow.collect { println(it) }
    }
    mongoClient.close()
}
